Reorganize documentation for MathML global attributes

fred-wang opened the pull request:
- Add a new article for the
  [math-depth](https://w3c.github.io/mathml-core/#the-math-script-level-property)
  property, which is used to implement the global attribute `scriptlevel`
  in MathML Core.
- Remove all per-element references to
  [global attributes](https://w3c.github.io/mathml-core/#global-attributes)
  from MathML element pages as well as the  Gecko-specific notes for
  `mathvariant`. Instead, they are moved to dedicated pages about global
  attributes.
- Introduce a new `mathml/global_attributes/` directory, based on
  HTML's one. For now, most of the links for these attributes point to
  equivalent pages for HTML. New articles are created for MathML-specific
  global attributes.

Note that the specifications and browser compatibility data for these new
articles are currently missing.

#### Summary

Reorganize documentation for MathML global attributes, moving them to dedicated pages instead of duplicating on each MathML element article. This also introduces CSS `math-depth` article for the sake of documenting `scriptlevel`.

#### Motivation

In the past, some attributes could be applied to a large class of MathML elements but not necessarily to all of them. This has changed with MathML Core where these generic attributes are now clearly defined as [global attributes](https://w3c.github.io/mathml-core/#global-attributes). Additionally, many attributes have now been taken from HTML (such as `nonce` or `on*` or `data-*`) and need to be documented.

#### Supporting details

https://w3c.github.io/mathml-core/#global-attributes

#### Related issues
<!-- 🔨 If applicable, use "Fixes #XYZ" -->

#### Metadata
<!-- ✅ Check a box if applicable, like this: [x]

This PR…
-->
- [x] Adds a new document
- [x] Rewrites (or significantly expands) a document
- [x] Fixes a typo, bug, or other error

<!-- 👷‍♀️ After submitting, review the results of the "Checks" tab! -->

fred-wang committed:
Reorganize documentation for MathML global attributes

- Add a new article for the
  [math-depth](https://w3c.github.io/mathml-core/#the-math-script-level-property)
  property, which is used to implement the global attribute `scriptlevel`
  in MathML Core. Also mention font-size: math.
- Remove all per-element references to
  [global attributes](https://w3c.github.io/mathml-core/#global-attributes)
  from MathML element pages as well as the  Gecko-specific notes for
  `mathvariant`. Instead, they are moved to dedicated pages about global
  attributes.
- Introduce a new `mathml/global_attributes/` directory, based on
  HTML's one. For now, most of the links for these attributes point to
  equivalent pages for HTML. New articles are created for MathML-specific
  global attributes.

Note that the specifications and browser compatibility data for these new
articles are currently missing.

teoli2003 commented:
Hi!

I have a generic question: what is the implementation status of MathML-Core?

fred-wang commented:
short answer: implemented in Chromium under a flag, and some alignment has been done on the Gecko/WebKit side.

estelle commented:
Do we have MathML page types? cc: @wbamberg?

estelle commented on files/en-us/web/mathml/global_attributes/displaystyle/index.md:
```diff
@@ -0,0 +1,30 @@
+{{MathMLRef("Global_attributes")}}
+
+The **`displaystyle`** [global attribute](/en-US/docs/Web/MathML/Global_attributes) is a boolean setting the [math-style](/en-US/docs/Web/CSS/math-style) of a MathML element:
+  - `true`, which means `normal`.
```
this should likely be in a ## Values section.

estelle commented on files/en-us/web/mathml/global_attributes/mathbackground/index.md:
```diff
@@ -0,0 +1,24 @@
+{{MathMLRef("Global_attributes")}}
+
+The **`mathbackground`** [global attribute](/en-US/docs/Web/MathML/Global_attributes) sets the [background-color](/en-US/docs/Web/CSS/background-color) of a MathML element.
+
```
an example of how it is used would be good and what value syntaxes are valid. Perhaps a link to color values.

estelle commented on files/en-us/web/mathml/global_attributes/mathcolor/index.md:
```diff
@@ -0,0 +1,24 @@
+{{MathMLRef("Global_attributes")}}
+
+The **`mathcolor`** [global attribute](/en-US/docs/Web/MathML/Global_attributes) sets the [color](/en-US/docs/Web/CSS/color) of a MathML element.
+
```
same comment as above. Should people be opting for CSS instead of these attributes? If yes, adding a note to not use this attribute would be good too

wbamberg commented:
> Do we have MathML page types? cc: @wbamberg?

According to https://github.com/mdn/content/issues/15539, although MathML does have specialised page types, it's not P1, so isn't slated to get page types defined yet. But I'd be happy for someone to define them of course.

fred-wang commented:
@estelle Thanks for the review. I believe I've addressed all the feedback.

I'm not sure I know what page types are, but ideally I'd like to have a sidebar similar to https://developer.mozilla.org/en-US/docs/Web/HTML where elements and global attributes are accessible from the sidebar. Probably out of the scope for this commit. For now, I added a link to global attributes from the MathML attributes page.

estelle commented on files/en-us/web/mathml/global_attributes/scriptlevel/index.md:
```diff
@@ -0,0 +1,31 @@
+---
+{{MathMLRef("Global_attributes")}}
+
+The **`scriptlevel`** [global attribute](/en-US/docs/Web/MathML/Global_attributes) sets the [math-depth](/en-US/docs/Web/CSS/math-depth) of a MathML element. If `<U>` is an unsigned [integer](/en-US/docs/Web/CSS/integer) (i.e. with prefix sign symbol removed) then the accepted values are:
```
```suggestion
The **`scriptlevel`** [global attribute](/en-US/docs/Web/MathML/Global_attributes) sets the [math-depth](/en-US/docs/Web/CSS/math-depth) of a MathML element. 

## Values

If `<U>` is an unsigned [integer](/en-US/docs/Web/CSS/integer) (i.e. with prefix sign symbol removed) then the accepted values are:
```

estelle replied:
likely needs more than just the header, but I don't know mathML at all.

estelle replied:
maybe an example too.

estelle commented on files/en-us/web/mathml/global_attributes/mathbackground/index.md:
```diff
@@ -0,0 +1,62 @@
+{{MathMLRef("Global_attributes")}}
+
+The **`mathbackground`** [global attribute](/en-US/docs/Web/MathML/Global_attributes) sets the [background-color](/en-US/docs/Web/CSS/background-color) of a MathML element.
+
```
```suggestion

> **Note:** Use CSS for styling MathML whenever possible. The `mathbackground` attribute should only be included for applications that are not CSS-aware and will be overridden by the CSS `background-color` property value, if set.

```

estelle commented on files/en-us/web/css/math-depth/index.md:
```diff
@@ -0,0 +1,67 @@
+
+## Browser compatibility
+
+{{Compat}}
```
If there are any other CSS properties that are math-related, a ## See also would be fantastic.

estelle commented on files/en-us/web/mathml/attribute/index.md:
```diff
@@ -7,7 +7,8 @@ tags:
 {{MathMLRef}}
 
-This is an alphabetical list of MathML attributes. More details for each attribute are available on particular [element pages](/en-US/docs/Web/MathML/Element).
+This is an alphabetical list of MathML attributes. More details for each attribute are available on particular [element pages](/en-US/docs/Web/MathML/Element)
```
```suggestion
This is an alphabetical list of MathML attributes. More details for each attribute are available on relevant [MathML element pages](/en-US/docs/Web/MathML/Element)
```

estelle commented on files/en-us/web/mathml/global_attributes/displaystyle/index.md:
```diff
@@ -0,0 +1,33 @@
+
+## Values
+
+- `true`, which means `normal`.
```
```suggestion
- `true`
  -: Sets the display style to `normal`.
```

estelle commented on files/en-us/web/mathml/global_attributes/displaystyle/index.md:
```diff
@@ -0,0 +1,33 @@
+## Values
+
+- `true`, which means `normal`.
+- `false`, which means `compact`.
```
```suggestion
- `false`
   -: Sets the display style to `compact`.
```

estelle commented on files/en-us/web/mathml/global_attributes/displaystyle/index.md:
```diff
@@ -0,0 +1,33 @@
+
+## See also
+
+- All [global attributes](/en-US/docs/Web/MathML/Global_attributes).
```
if there is a CSS styles that has similar effect, please add it.

estelle commented on files/en-us/web/mathml/global_attributes/href/index.md:
```diff
@@ -0,0 +1,30 @@
+{{MathMLRef("Global_attributes")}}
+
+The **`href`** [global attribute](/en-US/docs/Web/MathML/Global_attributes) creates a hyperlink on the MathML element pointing to the specified URL.
+
```
Can we include an example or two?

## Examples

```html
<math href="x">
<maction href="y">
````

Per below, href is valid on all elements, so including a non-math element helps impress that note from the next section.

estelle commented on files/en-us/web/mathml/global_attributes/mathcolor/index.md:
```diff
@@ -0,0 +1,81 @@
+{{MathMLRef("Global_attributes")}}
+
+The **`mathcolor`** [global attribute](/en-US/docs/Web/MathML/Global_attributes) sets the [color](/en-US/docs/Web/CSS/color) of a MathML element.
+
```
```suggestion

> **Note:** Use CSS for styling MathML whenever possible. The `mathcolor` attribute should only be included for applications that are not CSS-aware and will be overridden by the CSS `color` property , if set.

```

estelle replied:
Does `mathcolor` impact border-color like CSS `color` does?

fred-wang replied:
Yes `mathcolor` sets the `color` property via a [presentational hint](https://html.spec.whatwg.org/#presentational-hints) so all the feature of `color` apply too. Do you mean we should be more explicit about it?  I only explicitly mentioned "fraction bars or radical symbols" below as they are specific to math.

estelle commented on files/en-us/web/mathml/global_attributes/mathsize/index.md:
```diff
@@ -0,0 +1,24 @@
+{{MathMLRef("Global_attributes")}}
+
+The **`mathsize`** [global attribute](/en-US/docs/Web/MathML/Global_attributes) sets the [font-size](/en-US/docs/Web/CSS/font-size) of a MathML element.
+
```
values and example would be good.

estelle commented on files/en-us/web/mathml/global_attributes/mathvariant/index.md:
```diff
@@ -0,0 +1,53 @@
+---
+{{MathMLRef("Global_attributes")}}
+
+The **`mathvariant`** [global attribute](/en-US/docs/Web/MathML/Global_attributes)  attribute defines logical classes of token elements. Each class provides a collection of typographically-related symbolic tokens with specific meaning within a given mathematical expression. For `mathvariant` values other than `normal`, this is done by using glyphs of [Mathematical Alphanumeric Symbols](https://en ...
```
Is there a way to make this more understandable for the newbie? An example and value might help, as might a screen shot maybe?

estelle reviewed (commented):
a few more spots could use values and examples. I have a few questions too which I had reading this, so think addressing those may help the newbie (like me.)

Thanks!

wbamberg commented:
> @estelle Thanks for the review. I believe I've addressed all the feedback.
> 
> I'm not sure I know what page types are, but ideally I'd like to have a sidebar similar to https://developer.mozilla.org/en-US/docs/Web/HTML where elements and global attributes are accessible from the sidebar. Probably out of the scope for this commit. For now, I added a link to global attributes from the MathML attributes page.

Page types are just a piece of metadata, represented as a `"page-type"` front matter key, which says what kind of page something is: for example if it documents a WebAPI instance method, or an HTML element, or a CSS property. We will use them to build sidebars and also to define the kind of content a given page ought to include. I guess MathML would at least have `mathml-element` and `mathml-attribute`.

teoli2003 commented:
> > @estelle Thanks for the review. I believe I've addressed all the feedback.
> > I'm not sure I know what page types are, but ideally I'd like to have a sidebar similar to https://developer.mozilla.org/en-US/docs/Web/HTML where elements and global attributes are accessible from the sidebar. Probably out of the scope for this commit. For now, I added a link to global attributes from the MathML attributes page.
> 
> Page types are just a piece of metadata, represented as a `"page-type"` front matter key, which says what kind of page something is: for example if it documents a WebAPI instance method, or an HTML element, or a CSS property. We will use them to build sidebars and also to define the kind of content a given page ought to include. I guess MathML would at least have `mathml-element` and `mathml-attribute`.

Yes, I would have:
- `mathml-element`
- `mathml-global-attribute`
- `mathml-attribute`

And the two common to every area:
- `guide`
- `landing`

fred-wang commented:
@estelle Thanks. I believe this is ready for another round of review!

teoli2003 commented on files/en-us/web/mathml/global_attributes/mathsize/index.md:
```diff
@@ -0,0 +1,57 @@
+- {{cssxref("&lt;percentage&gt;")}}
+  - : A positive {{cssxref("&lt;percentage&gt;")}} value, relative to the parent element's font size.
+
+## Notes on MathML versions
+
+* In MathML 3 and earlier versions, keywords `small`, `normal`, and `big` as
+  well as the MathML3-specific syntax for lengths were supported.
+  Since MathML Core, the syntax matches CSS
+  {{cssxref("&lt;length-percentage&gt;")}} values.
+
+* This attribute was designed for MathML applications that are not CSS-aware.
+  Since MathML Core, use of equivalent CSS is recommended instead.
+
+
+## Specifications
+
+{{Specifications}}
```
```suggestion
## Notes on MathML versions




## Specifications

{{Specifications}}

- In MathML 3 and earlier versions, keywords `small`, `normal`, and `big` as
  well as the MathML3-specific syntax for lengths was supported.
  Since MathML Core, the syntax matches CSS
  {{cssxref("&lt;length-percentage&gt;")}} values.

* This attribute was designed for MathML applications that are not CSS-aware.
  Since MathML Core, the use of equivalent CSS is recommended instead.
```

teoli2003 commented on files/en-us/web/mathml/global_attributes/mathvariant/index.md:
```diff
@@ -0,0 +1,118 @@
+
+The **`mathvariant`** [global attribute](/en-US/docs/Web/MathML/Global_attributes)  attribute sets a logical class for textual elements, which is visually
+distinguished by using special [Mathematical Alphanumeric Symbols](https://en.wikipedia.org/wiki/Mathematical_Alphanumeric_Symbols).
+With the exception of
+[`mi`](/en-US/docs/Web/MathML/Element/mi) elements with a single character, which
+are by convention italic, no special classes are used by default.
```
NIT:
```suggestion
With the exception of [`mi`](/en-US/docs/Web/MathML/Element/mi) elements with a single character,
which are by convention italic, no special classes are used by default.
```

teoli2003 commented on files/en-us/web/mathml/global_attributes/mathvariant/index.md:
```diff
@@ -0,0 +1,118 @@
+   -: Try and use stretched characters e.g. "𞹢".
+
+
+## Notes on MathML versions
+
+* `mathvariant` values other than `normal` were designed for backward compatibility. Authors are encouraged to instead directly use the corresponding Unicode characters from the [Mathematical Alphanumeric Symbols](https://en.wikipedia.org/wiki/Mathematical_Alphanumeric_Symbols).
```
```suggestion
## Notes on MathML versions

> **Note:** `mathvariant` values other than `normal` were designed for backward compatibility. Authors are encouraged to instead directly use the corresponding Unicode characters from the [Mathematical Alphanumeric Symbols](https://en.wikipedia.org/wiki/Mathematical_Alphanumeric_Symbols).
```

teoli2003 commented on files/en-us/web/mathml/global_attributes/mathvariant/index.md:
```diff
@@ -0,0 +1,118 @@
+
+* `mathvariant` values other than `normal` were designed for backward compatibility. Authors are encouraged to instead directly use the corresponding Unicode characters from the [Mathematical Alphanumeric Symbols](https://en.wikipedia.org/wiki/Mathematical_Alphanumeric_Symbols).
+
+* MathML Core relies on `text-transform` to implement the `mathvariant` attribute, but implementations of older MathML versions may follow a different approach to implement similar behavior.
+
```
Same question as before: this looks like an implementation detail with no impact on developers.

fred-wang replied:
Yes, same answer.

fred-wang replied:
Handled in https://github.com/mdn/content/pull/18599

teoli2003 commented on files/en-us/web/mathml/global_attributes/mathvariant/index.md:
```diff
@@ -0,0 +1,118 @@
+- All [global attributes](/en-US/docs/Web/MathML/Global_attributes).
+- {{cssxref("text-transform")}}
+
+## Gecko-specific notes
+
+- Some `mathvariant` values are only implemented starting with Gecko 28.0 (Firefox 28.0 / Thunderbird 28.0 / SeaMonkey 2.25) and require appropriate [math fonts](/en-US/docs/Web/MathML/Fonts).
```
```suggestion
```

teoli2003 commented on files/en-us/web/mathml/global_attributes/mathvariant/index.md:
```diff
@@ -0,0 +1,118 @@
+## Browser compatibility
+
+{{Compat}}
+
```
Taken from below:

```suggestion
On Firefox, some `mathvariant` values are only implemented starting with Firefox 28 and require appropriate [math fonts](/en-US/docs/Web/MathML/Fonts).
```

teoli2003 commented on files/en-us/web/mathml/global_attributes/scriptlevel/index.md:
```diff
@@ -0,0 +1,72 @@
+      -: Sets the `math-depth` to value `add(-<U>)`. This will scale up
+        `font-size` on the element `<U>` times.
+
+## Notes on MathML versions
+
+* MathML Core relies on `math-depth` to implement the `scriptlevel` attribute, but implementations of older MathML versions may follow a different approach to implement similar side-effect on the `font-size` within mathematical formulas.
```
Here again, isn't this an implementation detail?

fred-wang replied:
Same.

fred-wang replied:
Handled in https://github.com/mdn/content/pull/18599

teoli2003 reviewed (changes requested):
Thanks a lot, @fred-wang . I had a few comments/changes, but this is very close to being ready to merge.

fred-wang committed:
Update files/en-us/web/css/math-depth/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

fred-wang committed:
Update files/en-us/web/css/math-depth/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

fred-wang commented on files/en-us/web/css/math-depth/index.md:
```diff
@@ -0,0 +1,72 @@
+/* Keyword values */
+math-depth: auto-add;
+
+/*
```
Right, it seems I forgot to put the examples.

fred-wang committed:
Update files/en-us/web/css/math-depth/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

fred-wang committed:
comment syntax examples

fred-wang commented on files/en-us/web/css/math-depth/index.md:
```diff
@@ -0,0 +1,72 @@
+
+- `auto-add`
+  - : Set to the inherited `math-depth` plus 1 when inherited [math-style](/en-US/docs/Web/CSS/math-style) is `compact`.
+- `add(<integer>)`
```
I don't understand what you mean here? How do we know there is a "specified integer" if it's removed from the syntax?

fred-wang committed:
link to <integer> article

fred-wang commented on files/en-us/web/css/math-depth/index.md:
```diff
@@ -0,0 +1,72 @@
+  - : Set to the inherited `math-depth` plus 1 when inherited [math-style](/en-US/docs/Web/CSS/math-style) is `compact`.
+- `add(<integer>)`
+  - : Set to the inherited `math-depth` plus specified integer.
+- `<integer>`
```
Done.

fred-wang committed:
Update files/en-us/web/css/math-depth/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

fred-wang commented on files/en-us/web/mathml/element/maction/index.md:
```diff
@@ -24,16 +24,6 @@ The MathML **`<maction>`** element provides a possibility to bind actions to (su
-  - : The background color. You can use `#rgb`, `#rrggbb` and [HTML color names](/en-US/docs/Web/CSS/color_value#color_keywords).
-- mathcolor
-  - : The text color and also the fraction line color. You can use `#rgb`, `#rrggbb` and [HTML color names](/en-US/docs/Web/CSS/color_value#color_keywords).
 - selection
```
OK, but I think it should be a separate thing when we rewrite the element pages.

fred-wang replied:
Still TODO.

fred-wang replied:
https://github.com/mdn/content/pull/18629

fred-wang committed:
s/MDNSidebar/CSSRef

fred-wang commented on files/en-us/web/mathml/element/mfenced/index.md:
```diff
@@ -15,16 +15,8 @@ The deprecated MathML `<mfenced>` element used to provide the possibility to add
 
 ## Attributes
```
Yes, the plan was to do more changes to element article in separate PR as it's already big. Although it's true we would be temporarily losing that info...

fred-wang commented on files/en-us/web/mathml/element/mi/index.md:
```diff
@@ -12,105 +12,6 @@ browser-compat: mathml.elements.mi
 
 The MathML `<mi>` element indicates that the content should be rendered as an **identifier** such as function names, variables or symbolic constants. You can also have arbitrary text in it to mark up terms.
 
-## Attributes
```
Yes no more attributes.

fred-wang commented on files/en-us/web/mathml/element/mi/index.md:
````diff
@@ -12,105 +12,6 @@ browser-compat: mathml.elements.mi
-
 ## Examples
 
 ```html
````
`mathvariant` is a global attribute, I guess we can remove the example indeed. Can we do that in an "update all mathml elements" PR instead?

fred-wang committed:
Update files/en-us/web/mathml/element/mo/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

fred-wang commented on files/en-us/web/mathml/element/mstyle/index.md:
```diff
@@ -14,14 +14,8 @@ The MathML `<mstyle>` element is used change the style of its children. It accep
 
 ## Attributes
```
Yes, both of these attributes are now global. Checking this page, it needs a lot of refresh, so let's do that in separate PR?

fred-wang replied:
Article rewritten in https://github.com/mdn/content/pull/18639

fred-wang commented on files/en-us/web/mathml/global_attributes/displaystyle/index.md:
```diff
@@ -0,0 +1,63 @@
+- `false`
+   -: Sets the display style to `compact`.
+
+## Notes on MathML versions
+
+* MathML Core relies on `math-style` to implement the `displaystyle` attribute, but implementations of older MathML versions may follow a different approach to implement similar behavior.
```
So the idea is that browsers will expose this via a CSS property and so developers can potentially directly get/set that property via other APIs... which won't work with legacy engines.

fred-wang replied:
Done in https://github.com/mdn/content/pull/18599

fred-wang committed:
Update files/en-us/web/mathml/global_attributes/href/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

fred-wang committed:
Update files/en-us/web/mathml/global_attributes/mathvariant/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

fred-wang committed:
Update files/en-us/web/mathml/global_attributes/mathvariant/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

fred-wang committed:
Update files/en-us/web/mathml/global_attributes/mathbackground/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

fred-wang committed:
Update files/en-us/web/mathml/global_attributes/mathcolor/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

fred-wang committed:
Update files/en-us/web/mathml/global_attributes/mathsize/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

fred-wang committed:
Update files/en-us/web/mathml/global_attributes/mathvariant/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

teoli2003 commented:
Let's merge this!

fred-wang commented on files/en-us/web/css/math-depth/index.md:
```diff
@@ -0,0 +1,72 @@
+
+## Formal definition
+
+{{cssinfo}}
```
Still TODO:
https://developer.mozilla.org/en-US/docs/Web/CSS/math-depth

fred-wang replied:
This one (and the compat/spec data) will hopefully be fixed by https://github.com/mdn/browser-compat-data/pull/17071

fred-wang commented on files/en-us/web/css/math-depth/index.md:
```diff
@@ -0,0 +1,72 @@
+
+## Formal syntax
+
+{{CSSSyntax}}
```
Still TODO:
https://developer.mozilla.org/en-US/docs/Web/CSS/math-depth

fred-wang replied:
This does not work for math-style either. I've opened issue https://github.com/mdn/content/issues/18641

I think we can now forget about this PR, all the review comments have been addressed or moved to other places.
